lib/gcloud/bigquery/dataset.rb

```diff
@@ -0,0 +1,161 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "json"
+require "gcloud/bigquery/errors"
+require "gcloud/bigquery/table"
+require "gcloud/bigquery/list"
+
+module Gcloud
+  module Bigquery
+    ##
+    # Represents a Dataset.
+    class Dataset
+      ##
+      # The Connection object.
+      attr_accessor :connection #:nodoc:
+
+      ##
+      # The Google API Client object.
+      attr_accessor :gapi #:nodoc:
+
+      ##
+      # Create an empty Dataset object.
+      def initialize #:nodoc:
+        @connection = nil
+        @gapi = {}
+      end
+
+      ##
+      # A unique ID for this dataset, without the project name.
+      # The ID must contain only letters (a-z, A-Z), numbers (0-9),
+      # or underscores (_). The maximum length is 1,024 characters.
+      def dataset_id
+        @gapi["datasetReference"]["datasetId"]
+      end
+
+      ##
+      # The ID of the project containing this dataset.
+      def project_id
+        @gapi["datasetReference"]["projectId"]
+      end
+
+      ##
+      # A descriptive name for the dataset.
+      def name
+        @gapi["friendlyName"]
+      end
+
+      ##
+      # A user-friendly description of the dataset.
+      def description
+        @gapi["description"]
+      end
+
+      ##
+      # The default lifetime of all tables in the dataset, in milliseconds.
+      def default_expiration
+        @gapi["defaultTableExpirationMs"]
+      end
+
+      ##
+      # The time when this dataset was created.
+      def created_at
+        return nil if @gapi["creationTime"].nil?
+        Time.at(@gapi["creationTime"] / 1000.0)
+      end
+
+      ##
+      # The date when this dataset or any of its tables was last modified.
+      def modified_at
+        return nil if @gapi["lastModifiedTime"].nil?
+        Time.at(@gapi["lastModifiedTime"] / 1000.0)
+      end
+
+      ##
+      # Permenently deletes the dataset.
+      # The dataset must be empty.
+      #
+      #   dataset.delete
+      def delete options = {}
+        ensure_connection!
+        resp = connection.delete_dataset dataset_id, options
+        if resp.success?
+          true
+        else
+          ApiError.from_response(resp)
+        end
+      end
+
+      ##
+      # Creates a table on a given dataset for a given subscriber.
+      #
+      # If the name is not provided in the request, the server will assign a
+      # random name for this table on the same project as the dataset.
+      def create_table options = {}
+        ensure_connection!
+        resp = connection.insert_table dataset_id, options
+        if resp.success?
+          Table.from_gapi resp.data, connection
+        else
+          # TODO: Handle ALREADY_EXISTS and NOT_FOUND
+          fail ApiError.from_response(resp)
+        end
+      end
+
+      ##
+      # Retrieves a table by name.
+      def table table_name
```

Comment:
should this be "table" or "get_table" ? Thought "get" was our way of saying "this will make an API call" ?